docs(samples): Query profiling feature samples

danieljbruce opened the pull request:
**Changes:**

In _samples/queryFilterOr.js_, the sort isn't working properly. This results in a flakey test for OR queries because runQuery doesn't guarantee an order for results and relies the sort so that the results are in the right order for the assert statement. The sort function is fixed so it now works properly.

Samples files _samples/queryProfileExplain.js_, _samples/queryProfileExplainAggregation.js_, _samples/queryProfileExplainAnalyze.js_, _samples/queryProfileExplainAnalyzeAggregation.js_ are added which contain the code samples that execute runQuery and runAggregationQuery for both query planning and query profiling. The filenames match the filenames in java-datastore.

In _samples/test/queryFilterOr.test_ we pull out the code that sets the database to a specific state into a _samples/test/helpers/populate-data.js_ file because we want to reuse this code to set the database to a specific state for the query profiling samples as well.

In _samples/test/queryProfiling.test.js_ we add samples tests that run each of the 4 files with query profiling code samples and compare the output in each sample to the expected output for each code sample.

danieljbruce committed:
Add the right files from the preview branch

Copy paste over the files from the preview branch

danieljbruce committed:
Thread Query mode through the library

Thread the query mode through the client library. Allow calls to get made using the query mode.

danieljbruce committed:
Create a basic workflow for testing profiling grpc

This code allows basic query profiling to work and creates a basic code snippet that can be used to make it work.

danieljbruce committed:
This test makes a call to the mock server

This will be useful for debugging grpc issues.

danieljbruce committed:
Now the test works with the mock server in repo

This mock server will be super useful for ensuring that the right data gets transmitted in private preview so that we can confirm the client library is working correctly.

danieljbruce committed:
Delete js file

We use the ts file instead.

danieljbruce committed:
Try explain

See if Explain gets passed along to the mock server.

danieljbruce committed:
Update protos.json

Protos.json is needed to pass the plan along.

danieljbruce committed:
Try an aggregate query

danieljbruce committed:
Add test to catch breaking change to return type

Make sure changes don’t break the use of this callback by adding this test.

danieljbruce committed:
Try redefining RunQueryCallback

Define it with a more general data structure.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into query-profiling-feature

# Conflicts:
#	protos/google/datastore/v1/query_profile.proto
#	protos/protos.d.ts
#	protos/protos.js
#	protos/protos.json
#	src/v1/datastore_client.ts

danieljbruce committed:
Revert Gapic changes to match head branch

The merge caused generated files to have a diff. The diff should not be in the PR.

danieljbruce committed:
Remove only

only should not be used here

danieljbruce committed:
Add data structures for return values

Data structures need to be added for the return values that contain stats.

danieljbruce committed:
Add plumbing to send stats info back

Add stats to info if it is available. If there are no results then end the stream and send the info back. This way, stats will always be sent in info if they are available and the program won ’t break if there are no results.

danieljbruce committed:
Set test to only that we intend to explore

danieljbruce committed:
Add a comment about stats

Explain what happens when the result set is empty. Just send the stats back.

danieljbruce committed:
Delete the mock server code

The mock server code was used for debugging and now we don’t need it since the service is working.

danieljbruce committed:
Remove calls to nightly

Calls to nightly don’t have a second database to work with. Regular calls work now so nightly calls are not necessary.

danieljbruce committed:
Introduce profiling tests again

Bring the query profiling tests back.

danieljbruce committed:
Revert "Remove calls to nightly"

This reverts commit 040d0a5718e913b1a5b31e37a13b0a5daa74f4a1.

danieljbruce committed:
Stats are optional

Stats do not necessarily come from the server.

danieljbruce committed:
Write some tests to test each mode

Each query profiling mode needs to be explored.

danieljbruce committed:
Add code for parsing the stats returned

Stats returned needs to be parsed by a special library that will remove the complexities of the Struct object.

danieljbruce committed:
Add dependencies necessary for parsing

A library is needed for parsing the Struct values. Add the libraries and use them.

danieljbruce committed:
Add assertions for the expected plan/stats

Expected plan and expected stats should be used in the tests. This ensures the tests check for the proper stats values.

danieljbruce committed:
Refactor info with stats build and add info to cb

Add a specific type for the runAggregationQuery callback so that it can now support the info parameter. In order to allow runAggregationQuery to make use of creating info, we also refactor info construction into a separate function.

danieljbruce committed:
Modify the parser for runAggregationQuery

The parser for runAggregationQuery should have a deterministic process for computing results.

danieljbruce committed:
Add asserts for the return values of runAggregate

Make sure that the entities and the plan are correct.

danieljbruce committed:
Complete tests for runQuery and aggregation query

The assertion checks for runQuery and runAggregationQuery should be done and they test each mode.

danieljbruce committed:
Add tests for Query and AggregateQuery

Tests should ensure run functions work properly for the run function on both of these objects.

danieljbruce committed:
Add initial transaction tests for runQuery

The runQuery tests have been added which get the right results. Next the right assert statements and info collection will be added to the tests.

danieljbruce committed:
Add checks on info to the transaction tests

Checks against info are needed to be sure that stats are passed back to the caller properly.

danieljbruce committed:
Fix tests for aggregate queries in transactions

Add tests for running aggregate queries inside transactions.

danieljbruce committed:
Ran linter, added all tests for runQueryStream

Added a temporary test for runQueryStream. Also ran the linter.

danieljbruce committed:
Change parsing of return values

Return values are going to look different for users. Change the code so that the parsing is done differently.

danieljbruce committed:
Reformat the info function

This function is more readable if we eliminate some of the tertiary arguments and complex logic for building the info object.

danieljbruce committed:
Change tests as a result of structure changes

The structure of the QueryInfo object is changed. Modify the tests to work with the new structure.

danieljbruce committed:
Use import and not require

import is better for catching compile time errors and is more customary.

danieljbruce committed:
Better spacing for imports

Change the spacing so that the imports are all in one place.

danieljbruce committed:
Introduce a single function for checking the execution stats. Make sure all the tests use this function. Pull out the run query plan and the run aggregation query plan.

danieljbruce committed:
Fix the tests so that they call the right fns

Add assertion checks to check the query plan against some expected value and make sure the right assertion checks are done for the right tests.

danieljbruce committed:
Finish the tests for the streaming call

Finish the tests for specifying no mode, specifying normal mode, EXPLAIN mode and EXPLAIN_ANALYZE mode. Make sure the tests  pass.

danieljbruce committed:
Delete code that will not be used anymore

There is a lot of boilerplate code that was needed for the streaming call. Get rid of it here.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into query-profiling-feature

danieljbruce committed:
Make changes to match new proto

Code change to use new proto was made so that code will compile.

danieljbruce committed:
Add Explain Metrics interface

Make slight change to withBeginTransaction so that code compiles under new structure. Also group plan and statistics under the new explainMetrics interface.

danieljbruce committed:
Remove bytesReturned from test

Proto should not be passing along bytesReturned anymore.

danieljbruce committed:
Fix system tests to use values matching new struct

danieljbruce committed:
Remove calls to nightly

Feature is now fully ready so nightly tests should not be done because the feature is expected to work in production.

danieljbruce committed:
Query profiling

Add test for runQuery. Send back plan summary and execution stats.

danieljbruce committed:
Add a test for runAggregationQuery

runAggregationQuery needs a unit test to validate request/return data.

danieljbruce committed:
Parameterize the query profiling tests

danieljbruce committed:
run the linter

danieljbruce committed:
Export Query Mode in index.ts

Query mode needs to be exported so that it can be accessed by the user.

danieljbruce committed:
Query profiling samples

Add query profiling samples for runQuery and runAggregationQuery.

danieljbruce committed:
Change data structure types to match return values

danieljbruce committed:
Remove TODO

danieljbruce committed:
remove import

danieljbruce committed:
delete the query profiling samples

danieljbruce committed:
Remove abstraction for RunQueryCallback

danieljbruce committed:
Change the comment to describe new data types

danieljbruce committed:
Remove TODO

danieljbruce committed:
Merge branch 'main' into query-profiling-feature

danieljbruce committed:
linting fixes

danieljbruce committed:
Update type to include runAggregationQuery

danieljbruce committed:
Put else back in

This change is actually simpler because it doesn’t introduce a let. It is also a much smaller diff.

danieljbruce committed:
mode is not needed in sharedQueryOptions

danieljbruce committed:
Revert "mode is not needed in sharedQueryOptions"

This reverts commit b8d0c63aed8ea0dec18efe61ae5022e853a70c78.

danieljbruce committed:
Merge branch 'query-profiling-feature' of https://github.com/danieljbruce/nodejs-datastore into query-profiling-feature-samples

# Conflicts:
#	samples/runAggregateQueryProfiling.js
#	samples/runQueryProfiling.js
#	samples/test/queryProfiling.test.js

danieljbruce committed:
Rearrange imports

rearrange the imports to simplify the diff.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into query-profiling-feature-samples

# Conflicts:
#	src/query.ts
#	src/request.ts
#	system-test/datastore.ts
#	test/index.ts

danieljbruce committed:
Undo old query profiling changes

danieljbruce committed:
Remove dependencies that are not needed

danieljbruce committed:
Pass analyze true in

danieljbruce committed:
Fix the flakey test - sort values

danieljbruce committed:
Move code out that is responsible for importing data into the datastore database in preparation for the samples.

Ensure query profiling output in the script matches what the sample tests expect.

danieljbruce committed:
Write the sample for aggregate queries

danieljbruce committed:
Add a description

danieljbruce committed:
Rename the sample files

danieljbruce committed:
Set files up so that they can be parsed

Rename the files too

danieljbruce committed:
Add some scripts for just the explain option

danieljbruce committed:
Merge branch 'main' into query-profiling-feature-samples

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
Remove logging of entities from the samples

kolea2 commented on samples/queryProfileExplain.js:
```diff
@@ -0,0 +1,38 @@
+
+// sample-metadata:
+//   title: Run query profiling (regular query)
+//   description: Run query profiling for a standard query and print query results
```
```suggestion
//   description: Run query explain for a standard query and print query explain metrics
```

kolea2 replied:
(here and in other places, please update)

kolea2 commented on samples/queryProfileExplain.js:
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+// sample-metadata:
+//   title: Run query profiling (regular query)
```
```suggestion
//   title: Run query explain (regular query)
```

kolea2 replied:
please update elsewhere, too

kolea2 commented on samples/queryProfileExplainAnalyze.js:
```diff
@@ -0,0 +1,44 @@
+  const [entities, info] = await datastore.runQuery(q, {
+    explainOptions: {analyze: true},
+  });
+  entities.sort((e1, e2) => {
```
why is sorting necessary?

danieljbruce replied:
Much like the [or query test](https://github.com/googleapis/nodejs-datastore/blob/b7ff5c86306f80d93f678a0638892c58a3b2088c/samples/test/queryFilterOr.test.js#L84) the entities are expected to be in a specific order for the test output. `runQuery` doesn't guarantee entities in a certain order so if we don't sort them then we end up with a flakey test.

kolea2 replied:
if this is only meant to assist in tests, I would prefer we change the test itself to assert on the contents of the results, ignoring the order.

danieljbruce replied:
Done

danieljbruce committed:
Update samples/queryProfileExplain.js

Co-authored-by: kolea2 <45548808+kolea2@users.noreply.github.com>

danieljbruce committed:
Update samples/queryProfileExplain.js

Co-authored-by: kolea2 <45548808+kolea2@users.noreply.github.com>

danieljbruce committed:
Update descriptions

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

kolea2 commented on samples/queryProfileExplainAggregation.js:
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+// sample-metadata:
+//   title: Run query profiling (aggregation query)
```
please fix to match other title (and in other places)

danieljbruce replied:
Done

kolea2 reviewed (dismissed):
looking good! A few small TODOs.

danieljbruce committed:
Don’t sort in the sample.

Make the tests indifferent about the order so that sorting isn’t required in the samples.

danieljbruce committed:
Fix titles

danieljbruce committed:
Add header

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

cindy-peng commented on samples/test/helpers/populate-data.js:
```diff
@@ -0,0 +1,56 @@
+  };
+
+  // Ensure the datastore database has no existing data.
+  const query = datastore.createQuery('Task');
```
Nit: does it make more sense to add these clean up to after() in L53?

danieljbruce replied:
No because then the samples tests won't work. This code is responsible for clearing the database and setting it to a specific state so that when the samples run the output is a predictable value.

danieljbruce committed:
Merge branch 'main' into query-profiling-feature-samples

bhshkh commented on samples/queryProfileExplain.js:
```diff
@@ -0,0 +1,38 @@
+  const [, info] = await datastore.runQuery(q, {
+    explainOptions: {analyze: false},
+  });
+  console.log(`info: ${Object.keys(info.explainMetrics)}`);
```
When I run this code, I'm seeing error: 
```
client.ts:32
      console.log(`info: ${Object.keys(info.explainMetrics)}`);
                                  ^

TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at quickstart (/usr/local/google/home/bahaaiman/Documents/cfdb-workspace-01/clients/node/nodejs-datastore-client/client.ts:32:35)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v20.14.0
```

What am I doing wrong in the code?
```
// Imports the Google Cloud client library
const { Datastore } = require('@google-cloud/datastore');

// Creates a client
const datastore = new Datastore({projectId: "my-project"});

async function quickstart() {
    // The Cloud Datastore key for the new entity
    const taskKey = datastore.key(['Task', 'task1']);

    // Prepares the new entity
    const task = {
        key: taskKey,
        data: {
            Description: 'Buy milk2',
            Done: false,
            Slice: [100, 200, 300]
        },
    };

    // Saves the entity
    await datastore.save(task);
    console.log(`Saved ${task.key.name}: ${task.data.Description}`);

    const query = datastore.createQuery('Task');
    const [entities, info] = await datastore.runQuery(query, {
        explainOptions: {analyze: true},
      });
      for (const entity of entities) {
        console.log(`Entity found: ${entity['description']}`);
      }
      console.log(`info: ${Object.keys(info.explainMetrics)}`);

}
quickstart();
```

bhshkh replied:
Checked with @danieljbruce offline. Updating the datastore version to 9.0.0 gets rid of the error
